src/main/scala/org/broadinstitute/hail/expr/AST.scala

```diff
@@ -227,8 +361,502 @@ sealed abstract class AST(pos: Position, subexprs: Array[AST] = Array.empty) {
       parseError(msg)
     subexprs.foreach(_.errorIf[T](msg))
   }
+
+  def compile(): CM[Code[AnyRef]]
+
+  def compileAggregator(): CMCodeCPS[AnyRef]
+
+  def run(ec: EvalContext): () => AnyRef = this.compile().run(ec)
+
+  def runAggregator(ec: EvalContext): CPS[Any] = {
+    val names = ec.st.toSeq
+      .sortBy { case (nanme, (i, _)) => i }
+      .map { case (name, _) => name }
+    val values = ec.a.asInstanceOf[mutable.ArrayBuffer[AnyRef]]
+
+    val idx = ec.a.length
+    val localA = ec.a
+    localA += null
```

Comment:
I feel dirty doing this, but I didn't take the time to think of a way to add a monadic action that is like `foreignFun` but which lets you pass the argument at a later time. I think I should be able to shim this into the initial environment (the first argument to the Function2Builder), but I haven't thought much about it.